pallets/pooled-staking/src/calls.rs

```diff
@@ -0,0 +1,459 @@
+
+            if Zero::is_zero(&value) {
+                continue;
+            }
+
+            match &operation {
+                PendingOperationKey::JoiningAutoCompounding {
+                    candidate,
+                    at_block,
+                } => {
+                    ensure!(
+                        T::JoiningRequestFilter::can_be_executed(&candidate, &delegator, *at_block),
+                        Error::<T>::RequestCannotBeExecuted(index as u16)
+                    );
+
+                    Self::execute_joining(
+                        candidate.clone(),
+                        delegator.clone(),
+                        TargetPool::AutoCompounding,
+                        Shares(value),
+                    )?;
+                }
+                PendingOperationKey::JoiningManualRewards {
+                    candidate,
+                    at_block,
+                } => {
+                    ensure!(
+                        T::JoiningRequestFilter::can_be_executed(&candidate, &delegator, *at_block),
+                        Error::<T>::RequestCannotBeExecuted(index as u16)
+                    );
+
+                    Self::execute_joining(
+                        candidate.clone(),
+                        delegator.clone(),
+                        TargetPool::ManualRewards,
+                        Shares(value),
+                    )?;
+                }
+                PendingOperationKey::Leaving {
+                    candidate,
+                    at_block,
+                } => {
+                    ensure!(
+                        T::LeavingRequestFilter::can_be_executed(&candidate, &delegator, *at_block),
+                        Error::<T>::RequestCannotBeExecuted(index as u16)
+                    );
+
+                    Self::execute_leaving(candidate.clone(), delegator.clone(), Shares(value))?;
+                }
+            }
+
+            PendingOperations::<T>::set(&delegator, &operation, Zero::zero());
+        }
+
+        Ok(().into())
+    }
+
+    fn execute_joining(
+        candidate: Candidate<T>,
+        delegator: Delegator<T>,
+        pool: TargetPool,
+        joining_shares: Shares<T>,
+    ) -> DispatchResultWithPostInfo {
+        // Convert joining shares into stake.
+        let stake = pools::Joining::<T>::sub_shares(&candidate, &delegator, joining_shares)?;
+        pools::Joining::<T>::decrease_hold(&candidate, &delegator, &stake)?;
+
+        // Convert stake into shares quantity.
+        let shares = match pool {
+            TargetPool::AutoCompounding => {
+                pools::AutoCompounding::<T>::stake_to_shares_or_init(&candidate, stake.clone())?
+            }
+            TargetPool::ManualRewards => {
+                pools::ManualRewards::<T>::stake_to_shares_or_init(&candidate, stake.clone())?
+            }
+        };
+
+        // If stake doesn't allow to get at least one share we release all the funds.
+        if Zero::is_zero(&shares.0) {
```

Comment:
If I understand correctly, this is because the stake was held before executing this operation (because this comes from a pending operation), and therefore if you cannot create at least one share you need to release it?

Reply:
When creating the request we cannot anticipate the value of a share due to rewards and slashing. So it could be that a user make a request for an amount that is now worth 1 share (+ dust) and that at execution is worth less than 1.